File: Data Structures & Algorithms/jump-game/submission-3.java
// class Solution {
//     public boolean canJump(int[] nums) {
//         int maxReachable = 0;
//         int n = nums.length;
//         for (int i = 0; i < n; i ++) {
//             if (i > maxReachable) return false;
//             maxReachable = Math.max(maxReachable, i + nums[i]);
//             if (maxReachable >= n - 1) {
//                 return true;
//             }
//         }
//         return true;
//     }
// }

class Solution {
    public boolean canJump(int[] nums) {
        int n = nums.length;
        int[] dp = new int[n];
        dp[0] = nums[0];
        for (int i = 1; i < n; i ++) {
            if (i <= dp[i-1]) {
                dp[i] = Math.max(dp[i-1], i + nums[i]);
            } else {
                dp[i] = dp[i-1];
            }
        }
        return dp[n-1] >= n-1;
    }
}

File: Data Structures & Algorithms/maximum-subarray/submission-3.py
# class Solution:
#     def maxSubArray(self, nums: List[int]) -> int:
#         res = float('-inf')
#         min_pre_sum = pre_sum = 0
#         for num in nums:
#             pre_sum += num
#             res = max(res, pre_sum - min_pre_sum)
#             min_pre_sum = min(min_pre_sum, pre_sum)
        
#         return res

class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        cur_sum = nums[0]
        max_sum = nums[0]
        for i in range(1, len(nums)):
            cur_sum = max(nums[i], cur_sum + nums[i])
            max_sum = max(cur_sum, max_sum)
        
        return max_sum

